### ISSUE TITLE
Nested sitemaps seem to 404 in Next.js 15

### ISSUE THREAD

**Issue by JamesSingleton** [vercel/next.js#72787]
### Link to the code that reproduces this issue

https://github.com/JamesSingleton/sitemap-issue

### To Reproduce

1. Start the app running `pnpm dev`
2. Go to `http://localhost:3000/resources/sitemap.xml/0` since in dev mode that's where it is supposed to live according to the [docs](https://nextjs.org/docs/app/api-reference/functions/generate-sitemaps)
3. You will get a 404
4. You can also do `pnpm build && pnpm start` and then go to `http://localhost:3000/resources/sitemap/0.xml` and you will get a 404 as well.

![Screenshot 2024-11-13 at 19 10 24](https://github.com/user-attachments/assets/bd2306ee-5e06-41fd-b7ec-acc01591389f)

### Current vs. Expected behavior

In Next 14 this worked just fine. However, after upgrading to Next 15 this no longer works.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 20.15.0
  npm: 10.7.0
  Yarn: 1.22.22
  pnpm: 9.13.0
Relevant Packages:
  next: 15.0.4-canary.11 // Latest available version is detected (15.0.4-canary.11).
  eslint-config-next: 15.0.3
  react: 19.0.0-rc-66855b96-20241106
  react-dom: 19.0.0-rc-66855b96-20241106
  typescript: 5.6.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local), Vercel (Deployed)

### Additional context

If you go to `localhost:3000/sitemap.xml` it will load that sitemap just fine
![Screenshot 2024-11-13 at 19 14 33](https://github.com/user-attachments/assets/94e3fe6a-e462-4680-a2b6-5126741f95ae)

I even tried moving the `sitemap.ts` to just `app/resources/sitemap.ts` and going to `http://localhost:3000/resources/sitemap.xml/0` still does not load anything

**Comment by cedricrabarijohn** [comment]
Same problem but using Next.js 14.2.12

**Comment by JamesSingleton** [comment]
> Same problem but using Next.js 14.2.12

I have an app on 14.2.15 and it works just fine. It’s when I moved to 15 it stopped working for me.

**Comment by cedricrabarijohn** [comment]
If you guys use Yoast for the sitemap generation for the wordpress application, I suggest you to go through this ressource : https://robertmarshall.dev/blog/use-yoast-sitemap-with-next-js-and-headless-wordpress . I think the approach is not bad at all.

**Comment by JamesSingleton** [comment]
> If you guys use Yoast for the sitemap generation for the wordpress application, I suggest you to go through this ressource : https://robertmarshall.dev/blog/use-yoast-sitemap-with-next-js-and-headless-wordpress . I think the approach is not bad at all.

I don't use Wordpress, in the example I gave you can see it just doesn't work even with a static array returned.

**Comment by JamesSingleton** [comment]
@leerob I apologize for tagging you in this, but this feels like a major regression in Next 15 from Next 14, especially with multiple people experiencing it.

**Comment by heggemsnes** [comment]
Also experiencing this. Working around it by creating a sitemap.xml folder with a route.tsx that exports a GET the "old fashioned way" for now.

**Comment by swarnava** [comment]
Hi,

The docs is outdated current;y, since v15 the urls are the always prod ones, which is sitemap/0.xml. It’s more reasonable now which always have the .xml extension and identical between dev and prod.

We will update the docs shortly.

**Comment by JamesSingleton** [comment]
> Hi,
> 
> The docs is outdated current;y, since v15 the urls are the always prod ones, which is sitemap/0.xml. It’s more reasonable now which always have the .xml extension and identical between dev and prod.
> 
> We will update the docs shortly.

I don't think it's just a docs issue, even the prod URLs don't work. For example, https://www.coparesources.com/resources/sitemap/0.xml used to return properly in Next 14, however, now in Next 15 it's broken returning a 404.

**Comment by samcx** [comment]
@JamesSingleton we add a hash after `sitemap-<xyz>`—so if you build your project, you should see where the exact path is (accessible in `next dev` or `next build`).

```
├   ├ /resources/sitemap-1jkbc8/0.xml
├   ├ /resources/sitemap-1jkbc8/1.xml
├   ├ /resources/sitemap-1jkbc8/2.xml
├   └ /resources/sitemap-1jkbc8/3.xml
```

**Comment by JamesSingleton** [comment]
@samcx is that new? Is that documented somewhere? I didn't see that anywhere which would be an undocumented breaking change 👎🏼

**Comment by samcx** [comment]
> I didn't see that anywhere which would be an undocumented breaking change 👎🏼

Agreed, so that's on us! That's why we're updating the docs now.

> Also, does that hash change on every single build?

I think it's more like the CSS Modules where it doesn't change unless something else changes (in the process of figuring that out).

**Comment by JamesSingleton** [comment]
>I think it's more like the CSS Modules where it doesn't change unless something else changes (in the process of figuring that out).

Let me know what you find out. It's a little concerning we are roughly a month and a half out from the v15 release and this was just now _caught_ (probably not the right terminology). This is a major change in how users handle sitemaps, you're basically forcing everyone to now go into their Google Search Console, Bing, etc and update the sitemaps that browsers look for because of a change Next.js made and not one they deliberately made.

**Comment by samcx** [comment]
@JamesSingleton Thank you for the feedback <picture data-single-emoji=":blob_bowing:" title=":blob_bowing:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/blob_bowing/3e13f8dbc67a5ea6.png" alt=":blob_bowing:" align="absmiddle"></picture>

**Comment by huozhi** [comment]
Hi, Thanks for flagging the outdated description in the docs. The behavior is consistent between dev and prod now since v15 RC release, you should now visit `resources/sitemap/0.xml`rather than `resources/sitemap.xml/0`.

We mentioned it as a breaking change in the [v15 blog](https://nextjs.org/blog/next-15)) but might it didn't highlight well with examples.
> [Breaking] Remove .xml extension for dynamic sitemap routes and align sitemap URLs between development and production ([PR](https://github.com/vercel/next.js/pull/65507))

I filed a PR (https://github.com/vercel/next.js/pull/73547) to updated the docs

**Comment by JamesSingleton** [comment]
@huozhi  that doesn’t close this issue out. I get that now between local development and production the URLs are the same, that’s create.

My issue is that Next now adds a hash to the URL as @samcx  pointed out. That’s a BREAKING CHANGE compared to the behavior that was there in Next 14. Going from `/resources/sitemap/0.xml` in production with Next 15 to now having `/resources/sitemap-1jkbc8/0.xml` is breaking. Furthermore, not knowing what causes the hash to change is concerning. Users now have to go into wherever they have their sitemaps and update it. If that hash changes whenever a new item is added, that’s going to cause A LOT is headaches for users.

**Comment by huozhi** [comment]
Thanks for clarifying I removed the closing directive in the PR. We used to add all the hash to the url under the group, for instance og image as they could be conflicted with the actual url. As you could have `app/resources/opengraph-image.tsx` and `app/(group)/resources/opengraph-image.ts`. But we did have an exception for the sitemap as sitemaps can be conflicted. I'll take another round look into that

**Comment by huozhi** [comment]
Yea that's what we made in v14 where the sitemap urls should be easier to discover and always normalized so that users can submit to indexer easily. And og images can just have the duplicates since they're always inlined in the page rather than exposing url for users to discover.

I'll address the fix soon, I misread what the "breaking change" refering to before. Thanks for keep clarifying here

**Comment by JamesSingleton** [comment]
No problem! I realized after @swarnava comment that I failed to mention the prod URL in my description originally. And then when it was mentioned it was just a docs update to address the issue I realized that I hadn't gotten the actual prod breaking change issue across, especially since I didn't even know in v15 a hash was added until @samcx mentioned that as it wasn't documented anywhere either.
